### PR TITLE
sql: allow booleans in SET tracing and friendlier error message

### LINKED ISSUE #44244: Internal error setting tracing=true
Consider the following statement:

```sql
SET SESSION tracing=true; -- ERROR: internal error: expected string for set tracing argument, not *tree.DBool
```

Unexpectedly, the `SET` results in the following internal error:

```sql
ERROR: internal error: expected string for set tracing argument, not *tree.DBool
SQLSTATE: XX000
DETAIL: stack trace:
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1184: runSetTracing()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:1134: runObserverStatement()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor_exec.go:73: execStmt()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1283: execCmd()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:1212: run()
github.com/cockroachdb/cockroach/pkg/sql/conn_executor.go:451: ServeConn()
github.com/cockroachdb/cockroach/pkg/sql/pgwire/conn.go:566: func1()
runtime/asm_amd64.s:1337: goexit()
```
I found this issue based on commit `c76ad970e73e606bc55a372d93f9d2d6acb32c9c`.

### PR DESCRIPTION
**Pull request by yuzefovich**
Previously, we would fail an assertion if a datum of the wrong type is
provided in `SET tracing` query (only strings were allowed). This resulted
in an internal error and printing out of the stack trace which can be
scary to users. This commit removes the assertion and makes it a regular
query error.

Also, booleans are now allowed as argument to `SET tracing`, and `true`
is mapped to `on` mode and `false` to `off`.

Fixes: #44244.

Release note (sql change, bug fix): Previously, CockroachDB would return
an internal error when using `SET tracing` with any type other than
string. Now it will return a regular query error. Additionally, boolean
arguments are now supported in `SET tracing`, and `true` is mapped to
`on` mode of tracing whereas `false` is mapped to `off`.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
Thanks for spotting it. Just a minor comment.

> Reviewed 2 of 2 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-:-LzHd88xAI7Apkg5VAKe:b-h6sr4n)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 1184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-LzHbSba1utk-XB9NhAi:-LzHbSba1utk-XB9NhAj:b-tebqyv) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b707346e48c018c278c003dd5592ddc9904d77/pkg/sql/conn_executor_exec.go#L1184)):*
> ```Go
> 		strVal, ok := v.(*tree.StrVal)
> 		if !ok {
> 			res.SetError(errors.Newf("expected string for set tracing argument, not %T", v))
> ```

I agree we need to fix this but this is not the way to go about it.
The main problem is that it's not good style to show something internal (%T) in a user-facing message.
The second problem is that it could be nice to allow booleans.

There are two ways to go:
- either the way of `(*planner) SetVar`
- or that of `(*planner)EvalAsOfTimestamp` (really `tree.EvalAsOfTimestamp`)

It's possible you can't use `analyzeExpr` /`Eval` like the first option, that's why I am mentioning the second.
(The reason why SET tracing is not an actual SetVar is that we want, I think, to be able to run it even when the txn is in error; in which case scalar evaluation is unavailable. But I'm not 100% sure. For the record, I prefer the SET (SetVar) semantics more.)

Long story short, get the expression, type switch on the result, and print a friendly error message.

Also the pg code for something that doesn't have the right type is `pgcode.Syntax`.
___
*[pkg/sql/logictest/testdata/logic_test/set, line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-LzHd4Rh2OnATTbX_5UR:-LzHd4Rh2OnATTbX_5US:b-4b7ry7) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b707346e48c018c278c003dd5592ddc9904d77/pkg/sql/logictest/testdata/logic_test/set#L305)):*
> ```
> SET ssl_renegotiation_limit = 123
> 
> statement error pq: expected string for set tracing argument, not \*tree.DBool
> ```

when you have a code, do `statement pgerror XXXX ....`

**Review by yuzefovich — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44260)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz and @yuzefovich)

___
*[pkg/sql/conn_executor_exec.go, line 1184 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-LzHbSba1utk-XB9NhAi:-LzIcBbAEWzE-TeIlHOQ:byb4r4z) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b707346e48c018c278c003dd5592ddc9904d77/pkg/sql/conn_executor_exec.go#L1184)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I agree we need to fix this but this is not the way to go about it.
The main problem is that it's not good style to show something internal (%T) in a user-facing message.
The second problem is that it could be nice to allow booleans.

There are two ways to go:
- either the way of `(*planner) SetVar`
- or that of `(*planner)EvalAsOfTimestamp` (really `tree.EvalAsOfTimestamp`)

It's possible you can't use `analyzeExpr` /`Eval` like the first option, that's why I am mentioning the second.
(The reason why SET tracing is not an actual SetVar is that we want, I think, to be able to run it even when the txn is in error; in which case scalar evaluation is unavailable. But I'm not 100% sure. For the record, I prefer the SET (SetVar) semantics more.)

Long story short, get the expression, type switch on the result, and print a friendly error message.

Also the pg code for something that doesn't have the right type is `pgcode.Syntax`.

</blockquote></details>

Thanks for the comment! I added the support for boolean arguments and changed the error to a regular query error.

In terms of `SetVar` semantics, I'm also not sure about the behavior there, so I'd rather stick with this least invasive approach.

**Review by knz — COMMENTED** [review]
LGTM modulo resolution of my comment below

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-:-LzIdSR45Knpai1BGrO0:b-n1ls5o)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @yuzefovich)

___
*[pkg/sql/logictest/testdata/logic_test/set, line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-LzHd4Rh2OnATTbX_5UR:-LzIdPrl6kQcLsgPzrhk:bkyl06x) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b707346e48c018c278c003dd5592ddc9904d77/pkg/sql/logictest/testdata/logic_test/set#L305)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

when you have a code, do `statement pgerror XXXX ....`
</blockquote></details>

can ou add an additional test that shows the syntax error

**Review by yuzefovich — COMMENTED** [review]
TFTR!

bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-:-LzJBOZZDfrmxz5eWuAs:b-oduykh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/sql/logictest/testdata/logic_test/set, line 305 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/44260#-LzHd4Rh2OnATTbX_5UR:-LzJBLkq-tT8-u3Jpe4g:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d2b707346e48c018c278c003dd5592ddc9904d77/pkg/sql/logictest/testdata/logic_test/set#L305)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

can ou add an additional test that shows the syntax error
</blockquote></details>

Done.

**Comment by yuzefovich** [comment]
Raphael, do you think we should backport it?

**Comment by knz** [comment]
Yes because we don't want this to continue spamming Sentry
-- 
Verstuurd vanaf mijn Android apparaat met K-9 Mail. Excuseer mijn beknoptheid.
